internal/encoding/decoder.go

```diff
@@ -0,0 +1,58 @@
+package encoding
+
+import (
+	"sync"
+)
+
+// Decoder decodes the contents of b into a v representation.
+// It's primarily used for decoding contents of a file into a map[string]interface{}.
+type Decoder interface {
+	Decode(b []byte, v interface{}) error
+}
+
+const (
+	ErrDecoderNotFound                = encodingError("decoder not found for this format")
+	ErrDecoderFormatAlreadyRegistered = encodingError("decoder already registered for this format")
+)
+
+// DecoderRegistry can choose an appropriate Decoder based on the provided format.
+type DecoderRegistry struct {
+	decoders map[string]Decoder
+
+	mu sync.RWMutex
+}
+
+// NewDecoderRegistry returns a new, initialized DecoderRegistry.
+func NewDecoderRegistry() *DecoderRegistry {
+	return &DecoderRegistry{
+		decoders: make(map[string]Decoder),
+	}
+}
+
+// RegisterDecoder registers a Decoder for a format.
+// Registering a Decoder for an already existing format is not supported.
+func (e *DecoderRegistry) RegisterDecoder(format string, enc Decoder) error {
```

Comment:
Would defining `type Key string` and replacing `format string` with `format Key` both here and in the `Decode()` method be appropriate? 

It's a trivial change, but cognitively it highlights the relationship between the parameters across the different methods while still allowing consumers to use strings.

Reply:
I'm not sure it'd be. What does `Key` mean? Format I understand.

Reply:
Perhaps `Key` is the wrong type name. The core of the suggestion is to just better highlight the relationship across methods rather than relying on variable names; perhaps it's unnecessary.